File: solver.py
import sys
import random
import math
from collections import defaultdict


import pandas as pd

def parse_excel_file(path):
    # Read all sheets from the excel file
    xls = pd.read_excel(path, sheet_name=None)
    out = {}
    for sheet_name, df in xls.items():
        # Clean up column names (strip whitespace)
        df.columns = df.columns.str.strip()
        # Convert to list of dicts
        # Replace NaN with None or empty string equivalent for compatibility
        data = df.where(pd.notnull(df), None).to_dict(orient='records')
        
        # Normalize sheet names to match what we expect (remove extra spaces or 'sheet' prefix if needed)
        # However, based on the txt file, expected keys are 'employees', 'vehicles', 'baseline', 'metadataa'
        # Let's clean the keys. 
        # The txt file had '#sheet employees', implying the sheet name is 'employees'.
        # Let's trust the excel sheet names are correct.
        out[sheet_name.strip()] = data
    
    # Handle the 'metadataa' vs 'metadata' possible typo if present in excel
    if 'metadata' in out and 'metadataa' not in out:
        out['metadataa'] = out['metadata']
        
    return out

def parse_test_file(path):
    sheets = {}
    cur = None
    with open(path, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            if line.startswith('#sheet'):
                cur = line.split()[1]
                sheets[cur] = []
                header = None
                continue
            if cur is None:
                continue
            # split by tabs
            parts = line.split('\t')
            if header is None:
                header = parts
                sheets[cur].append(header)
            else:
                sheets[cur].append(parts)
    # convert to list of dicts per sheet
    out = {}
    for name, rows in sheets.items():
        hdr = rows[0]
        data = []
        for r in rows[1:]:
            row = {k: v for k, v in zip(hdr, r)}
            data.append(row)
        out[name] = data
    return out


import datetime

def time_to_min(t):
    # t like '8:10' -> minutes since midnight
    if t is None or t == '':
        return None
    if isinstance(t, datetime.time):
        return t.hour * 60 + t.minute
    if isinstance(t, str):
        h, m = t.split(':')
        return int(h) * 60 + int(m)
    return t # fallback if already int or float minutes? No, usually not expected.


def haversine_km(lat1, lon1, lat2, lon2):
    R = 6371.0
    phi1 = math.radians(float(lat1))
    phi2 = math.radians(float(lat2))
    dphi = math.radians(float(lat2) - float(lat1))
    dlambda = math.radians(float(lon2) - float(lon1))
    a = math.sin(dphi/2)**2 + math.cos(phi1)*math.cos(phi2)*math.sin(dlambda/2)**2
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))


class GAPlanner:
    def __init__(self, employees, vehicles, baseline, metadata, pop_size=80, generations=200):
        self.employees = {e['employee_id']: e for e in employees}
        self.vehicles = {v['vehicle_id']: v for v in vehicles}
        self.baseline = {b['employee_id']: b for b in baseline}
        self.metadata = {m['key']: m['value'] for m in metadata}
        self.pop_size = pop_size
        self.generations = generations
        self.office_lat = list(self.employees.values())[0]['drop_lat']
        self.office_lng = list(self.employees.values())[0]['drop_lng']
        self.employee_ids = list(self.employees.keys())

    def initial_population(self):
        pop = []
        vehicle_ids = list(self.vehicles.keys())
        for _ in range(self.pop_size):
            # assign each employee to a random vehicle (may violate capacity)
            assign = {vid: [] for vid in vehicle_ids}
            for eid in self.employee_ids:
                vid = random.choice(vehicle_ids)
                assign[vid].append(eid)
            # randomize order within each vehicle
            for vid in vehicle_ids:
                random.shuffle(assign[vid])
            pop.append(assign)
        return pop

    def evaluate(self, assign):
        # compute total cost and total employee ride time
        total_cost = 0.0
        total_emp_time = 0.0
        penalty = 0.0

        baseline_total_cost = sum(float(self.baseline[e]['baseline_cost']) for e in self.employee_ids)
        baseline_total_time = sum(float(self.baseline[e]['baseline_time_min']) for e in self.employee_ids)

        for vid, e_list in assign.items():
            v = self.vehicles[vid]
            cap = int(v['capacity'])
            if len(e_list) > cap:
                penalty += 1000 * (len(e_list) - cap)
            # check sharing pref violations
            for eid in e_list:
                pref = self.employees[eid]['sharing_preference']
                if pref == 'single' and len(e_list) > 1:
                    penalty += 500
                if pref == 'double' and len(e_list) > 2:
                    penalty += 200
            # simulate route
            cur_lat = float(v['current_lat'])
            cur_lng = float(v['current_lng'])
            cur_time = time_to_min(v.get('available_from', '8:00'))
            total_dist = 0.0
            # pick ups
            onboard = []
            for eid in e_list:
                emp = self.employees[eid]
                plat = float(emp['pickup_lat'])
                plng = float(emp['pickup_lng'])
                d = haversine_km(cur_lat, cur_lng, plat, plng)
                total_dist += d
                travel_min = d / float(v['avg_speed_kmph']) * 60.0
                cur_time += travel_min
                # wait if earlier than earliest pickup
                if emp.get('earliest_pickup'):
                    ep = time_to_min(emp['earliest_pickup'])
                    if cur_time < ep:
                        cur_time = ep
                onboard.append(eid)
                cur_lat, cur_lng = plat, plng
            # go to office drop
            d = haversine_km(cur_lat, cur_lng, float(self.office_lat), float(self.office_lng))
            total_dist += d
            travel_min = d / float(v['avg_speed_kmph']) * 60.0
            cur_time += travel_min
            drop_time = cur_time
            # cost for vehicle
            total_cost += total_dist * float(v['cost_per_km'])
            # employee ride times (pickup->drop): approximate by drop_time - pickup_arrival
            # For simplicity, compute average pickup arrival times by re-simulating
            cur_lat = float(v['current_lat'])
            cur_lng = float(v['current_lng'])
            cur_time = time_to_min(v.get('available_from', '8:00'))
            for eid in e_list:
                emp = self.employees[eid]
                plat = float(emp['pickup_lat'])
                plng = float(emp['pickup_lng'])
                d = haversine_km(cur_lat, cur_lng, plat, plng)
                travel_min = d / float(v['avg_speed_kmph']) * 60.0
                cur_time += travel_min
                if emp.get('earliest_pickup'):
                    ep = time_to_min(emp['earliest_pickup'])
                    if cur_time < ep:
                        cur_time = ep
                pickup_arrival = cur_time
                # compute ride time until drop
                # remaining distance from pickup to office
                d2 = haversine_km(plat, plng, float(self.office_lat), float(self.office_lng))
                ride_min = d2 / float(v['avg_speed_kmph']) * 60.0
                emp_ride = ride_min
                total_emp_time += emp_ride
                # check latest_drop constraint
                if emp.get('latest_drop'):
                    ld = time_to_min(emp['latest_drop'])
                    if drop_time > ld + int(self.metadata.get('priority_1_max_delay_min', 0)):
                        # naive: use small extra slack parameter; big penalty
                        penalty += 1000
                cur_lat, cur_lng = plat, plng

        # normalized objective relative to baseline sums
        norm_cost = total_cost / (baseline_total_cost + 1e-6)
        norm_time = total_emp_time / (baseline_total_time + 1e-6)
        w_cost = float(self.metadata.get('objective_cost_weight', 0.6))
        w_time = float(self.metadata.get('objective_time_weight', 0.4))
        score = w_cost * norm_cost + w_time * norm_time + penalty / 10000.0
        return score, total_cost, total_emp_time, penalty

    def mutate(self, assign):
        # move a random employee to another random vehicle or shuffle order
        new = {k: v.copy() for k, v in assign.items()}
        if random.random() < 0.7:
            # reassign
            from_vid = random.choice(list(new.keys()))
            if not new[from_vid]:
                return new
            eid = random.choice(new[from_vid])
            new[from_vid].remove(eid)
            to_vid = random.choice(list(new.keys()))
            new[to_vid].append(eid)
        else:
            vid = random.choice(list(new.keys()))
            random.shuffle(new[vid])
        return new

    def crossover(self, a, b):
        # child takes some vehicles assignments from a and rest from b
        child = {}
        vids = list(self.vehicles.keys())
        cut = random.randint(1, max(1, len(vids)-1))
        left = set(random.sample(vids, cut))
        for vid in vids:
            child[vid] = a[vid].copy() if vid in left else b[vid].copy()
        # ensure all employees present exactly once: repair
        present = [eid for lst in child.values() for eid in lst]
        missing = [eid for eid in self.employee_ids if eid not in present]
        duplicates = [eid for eid in present if present.count(eid) > 1]
        # remove duplicates extra occurrences
        for eid in set(duplicates):
            count = present.count(eid)
            removed = count - 1
            for vid in vids:
                if removed <= 0:
                    break
                if eid in child[vid]:
                    child[vid].remove(eid)
                    removed -= 1
        # add missing randomly
        for eid in missing:
            vid = random.choice(vids)
            child[vid].append(eid)
        for vid in vids:
            random.shuffle(child[vid])
        return child

    def solve(self, callback=None):
        pop = self.initial_population()
        scored = []
        for p in pop:
            s = self.evaluate(p)[0]
            scored.append((s, p))

        for gen in range(self.generations):
            scored.sort(key=lambda x: x[0])
            newpop = [p for (_, p) in scored[:max(2, int(0.2 * len(scored)))]]
            # elitism kept
            while len(newpop) < self.pop_size:
                a = random.choice(scored)[1]
                b = random.choice(scored)[1]
                child = self.crossover(a, b)
                if random.random() < 0.3:
                    child = self.mutate(child)
                newpop.append(child)
            pop = newpop
            scored = [(self.evaluate(p)[0], p) for p in pop]
            
            if callback and gen % 5 == 0:
                best_gen_score, best_gen_assign = min(scored, key=lambda x: x[0])
                callback({
                    'generation': gen,
                    'score': best_gen_score,
                    'assignment': best_gen_assign
                })

        scored.sort(key=lambda x: x[0])
        best_score, best_assign = scored[0]
        score, total_cost, total_emp_time, penalty = self.evaluate(best_assign)
        return best_assign, score, total_cost, total_emp_time, penalty


def main(path):
    if path.endswith('.xlsx'):
        data = parse_excel_file(path)
    else:
        data = parse_test_file(path)
    
    employees = data.get('employees', [])
    vehicles = data.get('vehicles', [])
    baseline = data.get('baseline', [])
    # Try 'metadataa' then 'metadata'
    metadata = data.get('metadataa', data.get('metadata', []))
    
    if not employees or not vehicles:
        print('missing sheets in input (checked employees, vehicles)')
        print('Available keys:', list(data.keys()))
        return
    planner = GAPlanner(employees, vehicles, baseline, metadata, pop_size=80, generations=200)
    best_assign, score, total_cost, total_time, penalty = planner.solve()
    print('GA result:')
    print(f'Score: {score:.4f}, Total cost: {total_cost:.2f}, Total_emp_time_min: {total_time:.1f}, penalty:{penalty}')
    for vid, elist in best_assign.items():
        if elist:
            print(f'{vid}: {",".join(elist)}')


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print('Usage: python solver.py test_case_excel.txt')
    else:
        main(sys.argv[1])

File: solver.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <map>
#include <algorithm>
#include <random>
#include <fstream>
#include <sstream>
#include <iomanip>

using namespace std;

// data structures
struct Employee {
    string id;
    int int_id;
    double pickup_lat;
    double pickup_lng;
    int earliest_pickup; // minutes
    int latest_drop;     // minutes
    int sharing_pref;    // 0=any, 1=single, 2=double, 3=triple
    double baseline_cost;
    double baseline_time;
};

struct Vehicle {
    string id;
    int int_id;
    int capacity;
    double speed_kmph;
    double cost_per_km;
    double current_lat;
    double current_lng;
    int available_from; // minutes
};

struct Config {
    double cost_weight;
    double time_weight;
    int pop_size;
    int generations;
    double office_lat;
    double office_lng;
    // max delay constraint slacks
    int slack = 30; 
};

Config config;
vector<Employee> employees;
vector<Vehicle> vehicles;
double baseline_total_cost = 0;
double baseline_total_time = 0;

// Random engine
mt19937 rng;

// Global map for quick lookup if needed? indices are enough.
// We assign employees to vehicles.
// Solution representation: vector of vector of employee indices for each vehicle.
typedef vector<vector<int>> Solution;

double haversine_km(double lat1, double lon1, double lat2, double lon2) {
    const double R = 6371.0;
    double phi1 = lat1 * M_PI / 180.0;
    double phi2 = lat2 * M_PI / 180.0;
    double dphi = (lat2 - lat1) * M_PI / 180.0;
    double dlambda = (lon2 - lon1) * M_PI / 180.0;
    double a = sin(dphi/2.0) * sin(dphi/2.0) +
               cos(phi1) * cos(phi2) *
               sin(dlambda/2.0) * sin(dlambda/2.0);
    double c = 2.0 * atan2(sqrt(a), sqrt(1.0 - a));
    return R * c;
}

// Permutation based optimization for small routes
void optimize_route_sequence(int v_idx, vector<int>& route, double& out_dist, double& out_time, double& out_penalty) {
    if (route.empty()) {
        out_dist = 0; out_time = 0; out_penalty = 0;
        return;
    }

    const auto& v = vehicles[v_idx];
    
    // Sort initially by earliest pickup to have a decent start frame
    sort(route.begin(), route.end(), [](int a, int b){
        return employees[a].earliest_pickup < employees[b].earliest_pickup;
    });

    // If capacity is small <= 6, brute force permutations
    // If larger, stick to the sorted heuristic or Nearest Neighbor
    // V04 is a van with capacity 6. 6! = 720 iterations. totally fine in C++.
    
    double best_dist = 1e9;
    double best_penalty = 1e9;
    double best_total_emp_time = 1e9;
    vector<int> best_perm = route;
    
    bool try_permutations = (route.size() <= 6);
    // If not trying permutations, we just evaluate the heuristic sort order once.
    // To implement "just evaluate once", we can wrap the loop logic.
    
    // To support `next_permutation`, we need the vector sorted logic? 
    // No, standard `next_permutation` goes through lexicographical orders. 
    // We want ALL permutations. So start with sorted indices to be safe, or just run full cycle.
    // Actually `next_permutation` needs sorted to cover ALL.
    if (try_permutations) {
        sort(route.begin(), route.end()); 
    }

    do {
        double current_dist = 0;
        double current_penalty = 0;
        double current_total_emp_time = 0;
        
        double cur_lat = v.current_lat;
        double cur_lng = v.current_lng;
        double cur_time = v.available_from;
        
        vector<double> pickup_times;
        pickup_times.reserve(route.size());

        // Traverse route
        bool possible = true;
        for (int e_idx : route) {
            const auto& emp = employees[e_idx];
            double d = haversine_km(cur_lat, cur_lng, emp.pickup_lat, emp.pickup_lng);
            current_dist += d;
            
            double travel = (d / v.speed_kmph) * 60.0;
            cur_time += travel;
            
            // Earliest pickup wait (hard constraint logic usually, but here wait counts to time?)
            // Usually wait time doesn't cost driver distance, but consumes time.
            if (emp.earliest_pickup > 0 && cur_time < emp.earliest_pickup) {
                cur_time = emp.earliest_pickup;
            }
            
            pickup_times.push_back(cur_time);
            cur_lat = emp.pickup_lat;
            cur_lng = emp.pickup_lng;
        }
        
        // Go to office
        double d_office = haversine_km(cur_lat, cur_lng, config.office_lat, config.office_lng);
        current_dist += d_office;
        double travel_office = (d_office / v.speed_kmph) * 60.0;
        cur_time += travel_office;
        double drop_time = cur_time;

        // Check constraints & penalties
        for (size_t i = 0; i < route.size(); ++i) {
            int e_idx = route[i];
            const auto& emp = employees[e_idx];
            
            double ride_time = drop_time - pickup_times[i];
            current_total_emp_time += ride_time;
            
            if (emp.latest_drop > 0) {
                if (drop_time > emp.latest_drop + config.slack) {
                     current_penalty += 1000.0 + (drop_time - emp.latest_drop);
                }
            }
        }
        
        // Compare with best
        // Objective for sequencing: Minimize (Cost + Penalty). 
        // We often care less about ride time inside the sequence optimization unless it generates penalty.
        // Actually the global objective includes ride time.
        // Local Score = CostWeight * Dist * CostPerKm + TimeWeight * EmpTime + Penalty
        // We need normalized values strictly speaking, but for local optimization, raw weighted sum is fine.
        // Since normalization factors are constant, we can use effective weights.
        
        // Approx weights for local decision
        double local_score = (config.cost_weight * current_dist * v.cost_per_km) + 
                             (config.time_weight * current_total_emp_time) + 
                             current_penalty;

        // Check if better
        double current_best_score = (config.cost_weight * best_dist * v.cost_per_km) + 
                                    (config.time_weight * best_total_emp_time) + 
                                    best_penalty;

        if (local_score < current_best_score) {
            best_dist = current_dist;
            best_total_emp_time = current_total_emp_time;
            best_penalty = current_penalty;
            best_perm = route;
        }

    } while (try_permutations && next_permutation(route.begin(), route.end()));

    // Set the route to the best permutation found
    route = best_perm;
    out_dist = best_dist;
    out_time = best_total_emp_time;
    out_penalty = best_penalty;
}

double evaluate(Solution& sol, double& out_cost, double& out_time, double& out_penalty) {
    double total_cost = 0;
    double total_emp_time = 0;
    double penalty = 0;

    // To allow `evaluate` to update the solution inplace with best sequence, we passed sol by ref.
    // This is "Lamarckian" evolution - the improvements during evaluation are written back to the individual.
    
    for (size_t v_idx = 0; v_idx < sol.size(); ++v_idx) {
        auto& route = sol[v_idx]; // list of emp indices
        if (route.empty()) continue;

        const auto& v = vehicles[v_idx];
        
        // Capacity penalty
        if (route.size() > (size_t)v.capacity) {
            penalty += 1000.0 * (route.size() - v.capacity);
            // If over capacity, optimization might be slow (n! grows fast). 
            // In mutation/init we might exceed capacity locally.
            // If > 8, do NOT permute.
            // But we truncated logic to <=6 inside function.
        }

        // Sharing preference penalty
        for (int e_idx : route) {
            int pref = employees[e_idx].sharing_pref;
            if (pref == 1 && route.size() > 1) penalty += 500;
            if (pref == 2 && route.size() > 2) penalty += 200;
        }

        double d, t, p;
        // Optimize sequence locally!
        optimize_route_sequence(v_idx, route, d, t, p);
        
        total_cost += d * v.cost_per_km;
        total_emp_time += t;
        penalty += p;
    }

    out_cost = total_cost;
    out_time = total_emp_time;
    out_penalty = penalty;

    // Objective
    double norm_cost = total_cost / (baseline_total_cost + 1e-6);
    double norm_time = total_emp_time / (baseline_total_time + 1e-6);
    
    return config.cost_weight * norm_cost + config.time_weight * norm_time + penalty / 10000.0;
}

Solution create_random_solution() {
    Solution sol(vehicles.size());
    // Random assign
    for (size_t i = 0; i < employees.size(); ++i) {
        int v_idx = rng() % vehicles.size();
        sol[v_idx].push_back(i);
    }
    // Shuffle routes
    for (auto& route : sol) {
        shuffle(route.begin(), route.end(), rng);
    }
    return sol;
}

Solution mutate(const Solution& p) {
    Solution child = p;
    if (employees.empty() || vehicles.empty()) return child;

    double r = (double)rng() / rng.max();
    if (r < 0.7) {
        // move employee
        // Pick random vehicle source
        int v_from = rng() % vehicles.size();
        if (child[v_from].empty()) {
            // try to find one with emps
            for(int k=0; k<10; ++k) {
                v_from = rng() % vehicles.size();
                if(!child[v_from].empty()) break;
            }
        }
        
        if (!child[v_from].empty()) {
             int idx_in_route = rng() % child[v_from].size();
             int e_idx = child[v_from][idx_in_route];
             
             child[v_from].erase(child[v_from].begin() + idx_in_route);
             
             int v_to = rng() % vehicles.size();
             child[v_to].push_back(e_idx);
        }
    } else {
        // shuffle a route
        int v = rng() % vehicles.size();
        if (!child[v].empty()) {
            shuffle(child[v].begin(), child[v].end(), rng);
        }
    }
    return child;
}

Solution crossover(const Solution& A, const Solution& B) {
    // Single point crossover of vehicles?
    // Child takes vehicles 0..k from A, k+1..N from B
    // Then repair duplicates/missing
    Solution child(vehicles.size());
    int cut = rng() % max(1, (int)vehicles.size());
    
    for (int i = 0; i < cut; ++i) child[i] = A[i];
    for (size_t i = cut; i < vehicles.size(); ++i) child[i] = B[i];

    // Repair
    vector<int> counts(employees.size(), 0);
    for (const auto& route : child) {
         for (int e : route) counts[e]++;
    }

    vector<int> missing;
    for (size_t i = 0; i < employees.size(); ++i) {
        if (counts[i] == 0) missing.push_back(i);
    }

    for (size_t v = 0; v < vehicles.size(); ++v) {
        for (size_t i = 0; i < child[v].size(); ) {
            int e = child[v][i];
            if (counts[e] > 1) {
                // remove
                child[v].erase(child[v].begin() + i);
                counts[e]--;
            } else {
                i++;
            }
        }
    }
    
    // Add missing
    for (int e : missing) {
        int v = rng() % vehicles.size();
        child[v].push_back(e);
    }

    return child;
}

void print_status(int gen, double score, const Solution& sol, double c, double t, double p) {
    // Print JSON-like string
    // {"generation": 1, "score": 0.5, "stats": {"cost": 100, "time": 200, "penalty": 0}, "assignment": {"V01": ["E01", ...], ...}}
    cout << "{\"generation\": " << gen << ", \"score\": " << score 
         << ", \"stats\": {\"cost\": " << c << ", \"time\": " << t << ", \"penalty\": " << p << "}"
         << ", \"assignment\": {";
    
    bool firstV = true;
    for (size_t v=0; v<vehicles.size(); ++v) {
        if (!firstV) cout << ", ";
        firstV = false;
        cout << "\"" << vehicles[v].id << "\": [";
        for (size_t i=0; i<sol[v].size(); ++i) {
            if (i > 0) cout << ",";
            cout << "\"" << employees[sol[v][i]].id << "\"";
        }
        cout << "]";
    }
    cout << "}}" << endl;
}

int main(int argc, char** argv) {
    if (argc < 2) {
        cerr << "Usage: solver <input_file>" << endl;
        return 1;
    }

    rng.seed(time(0));

    // READ INPUT
    // Format assumed:
    // cost_weight time_weight pop_size generations
    // office_lat office_lng
    // num_employees
    // id pickup_lat pickup_lng earliest latest sharing base_cost base_time
    // ...
    // num_vehicles
    // id cap speed cost lat lng avail
    // ...

    ifstream fin(argv[1]);
    if (!fin.is_open()) {
        cerr << "Cannot open input file" << endl;
        return 1;
    }

    fin >> config.cost_weight >> config.time_weight >> config.pop_size >> config.generations;
    fin >> config.office_lat >> config.office_lng;

    int nE;
    fin >> nE;
    for (int i=0; i<nE; ++i) {
        Employee e;
        e.int_id = i;
        fin >> e.id >> e.pickup_lat >> e.pickup_lng >> e.earliest_pickup >> e.latest_drop 
            >> e.sharing_pref >> e.baseline_cost >> e.baseline_time;
        employees.push_back(e);
        baseline_total_cost += e.baseline_cost;
        baseline_total_time += e.baseline_time;
    }

    int nV;
    fin >> nV;
    for (int i=0; i<nV; ++i) {
        Vehicle v;
        v.int_id = i;
        fin >> v.id >> v.capacity >> v.speed_kmph >> v.cost_per_km 
            >> v.current_lat >> v.current_lng >> v.available_from;
        vehicles.push_back(v);
    }

    // GA Loop
    vector<pair<double, Solution>> population;
    for(int i=0; i<config.pop_size; ++i) {
         Solution s = create_random_solution();
         double c, t, p;
         double score = evaluate(s, c, t, p);
         population.push_back({score, s});
    }

    for (int gen = 0; gen <= config.generations; ++gen) {
        // Sort
        sort(population.begin(), population.end(), 
             [](const pair<double, Solution>& a, const pair<double, Solution>& b){
                 return a.first < b.first;
             });
        
        // Report
        if (gen % 5 == 0 || gen == config.generations) {
            double cc, tt, pp;
            evaluate(population[0].second, cc, tt, pp);
            print_status(gen, population[0].first, population[0].second, cc, tt, pp);
        }
        
        if (gen == config.generations) break;

        // Elitism: keep top 20%
        int elite_count = max(2, (int)(0.2 * config.pop_size));
        vector<pair<double, Solution>> next_pop;
        for(int i=0; i<elite_count; ++i) next_pop.push_back(population[i]);

        // Breed
        while(next_pop.size() < (size_t)config.pop_size) {
            // Tournament or random selection? Random top 50% is easy
            int idx1 = rng() % (config.pop_size / 2);
            int idx2 = rng() % (config.pop_size / 2);
            Solution child = crossover(population[idx1].second, population[idx2].second);
            if (((double)rng() / rng.max()) < 0.3) {
                 child = mutate(child);
            }
            double c, t, p;
            double score = evaluate(child, c, t, p);
            next_pop.push_back({score, child});
        }
        population = next_pop;
    }

    return 0;
}
